### DISCUSSION TITLE
Any way to add timecodes to the left and right sides of the Duration Bar OSD

### DISCUSSION THREAD

**Discussion by advice2020** [mpv-player/mpv#14952] [General]
Probably just as complicated as trying to modify the OSC but I was wondering if anyone knew if it was possible to add timecodes to the left and right side of the Duration Bar OSD, similar to how the OSC has them?
I wanted to see if I could move my Duration Bar OSD to bottom of screen (which there appears to be an option for already) but I do not see of a way to add timecodes.
The only thing I see is the ability to show timecodes when using "o" or "Shift+P" hotkeys, but not desired to have them show up at the top of screen while the bar is as the bottom.
Might also need to change the width of the Duration Bar OSD as well to fit the time code text.
This could allow Duration Bar OSD to take the place of the OSC, just wanted to try it out if it is possible.
Customizing graphical stuff can be really tricky in MPV.

Thank You

**Comment by advice2020** [comment]
I guess the task itself isn't that important but mainly just curious what kind of editing can be done to default OSD items provided by MPV.  OSD text labels / graphical items such as bars and icons.

Thank You

**Comment by Samillion** [comment]
Indeed it is possible. If you look at `osc.lua` as a reference, find the timecode layout. For example:

```lua
    lo = add_layout("tc_left")
    lo.geometry =
        {x = posX - pos_offsetX, y = bottomrowY, an = 4, w = 110, h = 18}
    lo.style = osc_styles.timecodes
```

`tc_left` meaning "Timecode left side". This should help you find the rest of the code that handles that particular layout, in this case it would be:

```lua
    -- tc_left (current pos)
    ne = new_element("tc_left", "button")

    ne.content = function ()
        if state.tc_ms then
            return (mp.get_property_osd("playback-time/full"))
        else
            return (mp.get_property_osd("playback-time"))
        end
    end
    ne.eventresponder["mbtn_left_up"] = function ()
        state.tc_ms = not state.tc_ms
        request_init()
    end
```

You can then compare that layout and its elements, then find the part for seekbar and simply adjust it (or repeat it) to the left and right.

**Comment by advice2020** [comment]
@Samillion 
Thank you for your response.
This post was regarding Duration OSD bar & time code text, not OSC.
Are you saying that the code for the OSD is in "osc.lua" file?
I am probably going to close this post because I do not think the idea I had for this is going to work that well.
But I am still interested in both topics: OSC modifications & OSD modifications.
I will keep this open based on your response, because I am still looking for an answer for if certain default OSDs provided by MPV can be modified, for example #14953.
But I was also looking on how to move items in the OSC, which was why I made that "geo" post you responded to, I will mention more about this in that post if interested.
Thank You

**Comment by Samillion** [comment]
Apologies, I misread and thought you were talking about OSC like the other posts.

Technically the answer is still apt, since OSD modification doesn't really exist, unless you consider `input.conf` with stuff like
```
RIGHT            osd-msg-bar seek  2 relative+exact
UP               osd-msg add volume 2
```

To truly customize the message output (ie: OSD message), you'll have to rely on OSC or a simple script that utilizes something like `show-text`

Though someone with more knowledge could enlighten us both on this, as I've not really ventured into OSD modification before. Just OSC.

**Comment by advice2020** [comment]
Ahh no problem, yeah I agree OSD has been a little confusing to me because it appears to be a general term for many different items, `show-text` is one that I have looked into which is the main one that allows users to customize text.
The OSD bars and text that are provided by default by MPV for volume and duration are the ones that do not appear that they can be modified.  I found that you can move the bars up and down if desired but do not think you can edit them or the text that appears at top of screen, probably have to turn the text off and replace with a `show-text` if desired, not sure yet.
I will leave this post open for a little longer to see if any others respond.
Thank you again
